### ISSUE TITLE
Package Addons on conda-forge

### ISSUE THREAD

**Issue by seanpmorgan** [tensorflow/addons#201]
This will likely come after we finalize our build automation and toolchain,  but wanted to start an issue for visibility. When the time comes we will enable anaconda installs using [conda-forge](https://conda-forge.org/#about).

cc @kyleabeauchamp

**Comment by kyleabeauchamp** [comment]
As a starting point, I made a WIP PR to get a conda package for TF2 nightly: https://github.com/conda-forge/tensorflow-feedstock/pull/68

**Comment by Smokrow** [comment]
Can I help with anything?

**Comment by seanpmorgan** [comment]
Hi @jjhelmus -- we're starting the process of packaging tensorflow/addons in conda and wanted to ask a few questions to see if you could guide us.

1. Is conda-forge the best channel for us?
1. It looks like there is a conda-forge/tensorflow and anaconda/tensorflow. Which dependency should we build against? (I think the conda-forge docs say to avoid external channel dependencies?)
1. Do you have an idea what the lag time is for tensorflow packages to show up on conda? (i.e. how long after tf2 release will we be able to build our conda package which depends on tf2?) It looks like conda-forge/tensorflow is still 1.13

**Comment by kyleabeauchamp** [comment]
1.  I'm 90% sure the answer is yes, but there will be issues with GPU support given the state of nvidia licensing (doesn't support OSS projects like conda-forge).  
2.  We're definitely in a stange place with wanting to support both CPU and GPU work.  I think this is a very good question that we need to resolve.
3.  The lag time for conda-forge is about an hour after the pull request is merged.  However, the review cycle can take up to a week.  

PS I still think it would be good to see if @jjhelmus agrees.

**Comment by xhochy** [comment]
As a conda-forge maintainer (not part of core though), I would answer:

1. Yes, it is. This where nearly all community-led conda package development happens except for specialized things like bioconda.
2. For CUDA in conda-forge and the legal problems see https://github.com/conda-forge/conda-forge.github.io/issues/63 and https://github.com/conda-forge/conda-forge.github.io/issues/687
3. Review cycles can take a week on complicated PRs but normally once CI has passed it's more around 1-2h during european and US daytime.

**Comment by jjhelmus** [comment]
1. conda-forge is the largest community-led packaging community and would make an ideal channel to use and contribute too (note that I am on the conda-forge core team so I might be bias).  Another option is to create your own channel and publish packages there.  
Some organizations like `pytorch` and `rapids` find this flexibility useful. Using your own channel will require you to setup the necessary infrastructure for producing packages and users may find it harder to find your packages than if they were on conda-forge.

2. This depends on how addons interacts with tensorflow.  If the interaction is via Python only then either channel should be fine.  If you make use of the C/C++ components then you will need to take care to match the ABI used by the channel.  
With the newer 1.13 builds conda-forge's linux-64 package is build from source using the conda compilers.  The osx-64 and win-64 packages are repackaged wheels. The packages in `defaults` have been built from source using the conda provided compilers for a number of versions.  
Only `defaults` provides a GPU package at this time.  The package name is the same, `tensorflow`, but the `tensorflow-gpu` metapackage can be used to select the GPU variant.
Please avoid using the `anaconda` channel.  It is a mixture of the two subchannels which make up `defaults` but without any subchannel priority which is needed for reliable solutions.

3. For conda-forge packages the time between a release and conda packages depends on entirely on volunteers.
For `defaults` the time depends on where the package is prioritized by Anaconda, Inc.  Typically packages are build within 2 weeks but this depends greatly on the difficulty of the build and other team priorities.  Given the large number of changes in TF 2.0 I would not be surprised if the lag for that release is longer than normal.

**Comment by seanpmorgan** [comment]
Thanks for all the feedback everyone. So to answer one question: Addons does indeed utilize the pre-compiled libtensorflow_framework.so for our C++ op linking. 

Just want verify my understanding from the above information is correct:
1) It seems like packaging in conda-forge will be ideal
2) We should link to `defaults` channel's `tensorflow` making sure we match the ABI that TF was compiled against
   * Should we also encourage users to install `cudatoolkit` if they want GPU support? Currently the plan is to link to tensorflow-gpu and use CPU fallback in case it is not available. In 2.1 these packages will be consolidated.
3) We are blocked on this issue until a TF2 full release is made and subsequent package is distributed on conda

**Comment by Smokrow** [comment]
Should we add like a "WiP" or "blocked" Label or something like that?

**Comment by seanpmorgan** [comment]
This is now unblocked as TF2 is on conda.

**Comment by mayou36** [comment]
Hi all, what's the status of this?

**Comment by seanpmorgan** [comment]
> Hi all, what's the status of this?

This is quite a bit more difficult than we originally suspected. A more complete investigation can be found for SIG IO who has the same goal:
https://github.com/tensorflow/io/issues/676

A more immediate fix would be the completion of https://github.com/tensorflow/community/pull/133 so that you can have a conda installed TF and a pip install TFA.

**Comment by mayou36** [comment]
Okey I see, many thanks. The problem is rather that we're having a downstream package [zfit](https://github.com/zfit/zfit) which we also like to publish on conda-forge (or rather, which we do, but adding TFA as a dependency breaks this)

**Comment by seanpmorgan** [comment]
Given our availability, we have decided to let the community package Addons on Conda. There are several channels:
https://anaconda.org/search?q=Addons+tensorflow

**Comment by mayou36** [comment]
Can we maybe reopen this? The problem with all these channels is that they are not in conda-forge. In practice, many people have conda-forge added to the default. Packages in conda-forge cannot just depend on other channels. So to have a downstream package in conda-forge, in reality, the upstream package is also needed in conda forge.

Since there are many channels, it would be good to let the community move it together. Can we maybe use this thread to organize this?

**Comment by h-vetinari** [comment]
For those interested, I opened a PR to add this to conda-forge: https://github.com/conda-forge/staged-recipes/pull/16888

**Comment by h-vetinari** [comment]
Hey all

I managed to get this to build and run the test suite, there's only a handful of tests that fail with some off-by-one errors. I'd really appreciate some input from the maintainers about these (and you'd be very welcome to co-maintain the conda-forge package! 😊)

```
=========================== short test summary info ============================
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-float32]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-float64]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-int32]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-uint8]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-int64]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-eager_mode-float16]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-float32]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-float64]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-int32]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-uint8]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-int64]
FAILED tensorflow_addons/image/tests/transform_ops_test.py::test_rotate_even[cpu-tf_function-float16]
==== 12 failed, 1872 passed, 572 skipped, 318 warnings in 396.17s (0:06:36) ====
```

Here's a [CI run](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=409878&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823) that has the full logs.

**Comment by jonas-eschle** [comment]
@seanpmorgan can we re-open this issue maybe?

**Comment by jonas-eschle** [comment]
@kyleabeauchamp could you maybe re-open this issue since there is progress by @h-vetinari?

**Comment by seanpmorgan** [comment]
Hi @h-vetinari thank you very much for your work on this. I've re-opened this issue seeing the progress you've made on the conda forge PR. Given our bandwidth for maintainers on this repo we'll be unable to support this ourselves but we're happy to onboard someone who would like to own this feature. We know this is a commonly requested feature. Please keep us updated on this progress.

**Comment by jonas-eschle** [comment]
@h-vetinari any news on this?

**Comment by h-vetinari** [comment]
> @h-vetinari any news on this?

Not so far, sorry. I didn't get any feedback from the maintainers here on the errors I noted above. But I can try to revive the PR and cross my fingers that those are gone with a newer version.

**Comment by h-vetinari** [comment]
Ah, sorry, I had misremembered the situation from https://github.com/conda-forge/staged-recipes/pull/16888. Letting bazel find a pre-installed (i.e. conda-forge-packaged) `tensorflow` looked to be extremely painful, and I didn't try to bash my head against this particular wall. Happy to receive help/guidance on this of course!

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
